iterators.tex

```diff
@@ -6327,19 +6327,26 @@
 
 \begin{itemdecl}
 template <class T>
-using IteratorType = decltype(begin(declval<T>()));
+  @\newtxt{requires requires (T\& t) \{}@
+    @\newtxt{\{ begin(t) \} -> Iterator; // not required to be equality preserving}@
+  @\newtxt{\}}@
+using IteratorType = decltype(begin(declval<T@\newtxt{\&}@>()));
 
 template <class T>
-using SentinelType = decltype(end(declval<T>()));
+  @\newtxt{requires requires (T\& t) \{}@
+    @\newtxt{typename IteratorType<T>;}@
+    @\newtxt{\{ end(t) \} -> Sentinel<IteratorType<T>{}>; // not required to be equality preserving}@
+  @\newtxt{\}}@
+using SentinelType = decltype(end(declval<T@\newtxt{\&}@>()));
```

Comment:
The motivation is to minimally ensure that `IteratorType<T>` is always an `Iterator`, and `SentinelType<T>` is always a `Sentinel`. We could theoretically require both `begin` and `end` for both associated types, or even require `Sentinel<end(t), begin(t)>` for both, I went with minimal requirements for no particular reason.